// react-chess/src/demo/demo.js
const React = require('react')
const Chess = require('../../src/react-chess')

require('./demo.css')

class Demo extends React.PureComponent {
  constructor(props) {
    super(props)

    this.state = {pieces: Chess.getDefaultLineup()}
    this.handleKeyUp = this.handleKeyUp.bind(this)
    this.handleMovePiece = this.handleMovePiece.bind(this)
    this.handleSelectPiece = this.handleSelectPiece.bind(this)
  }

  componentDidMount() {
    document.addEventListener('keyup', this.handleKeyUp, false)
  }

  componentWillUnmount() {
    document.removeEventListener('keyup', this.handleKeyUp, false)
  }

  handleKeyUp(evt) {
    const {selectedPiece, pieces} = this.state
    const isDeleteEvent = evt.key === 'Delete' || evt.key === 'Backspace'
    if (!isDeleteEvent) {
      return
    }

    evt.preventDefault()
    if (!selectedPiece) {
      return
    }

    this.setState({pieces: pieces.filter(piece => piece !== selectedPiece.notation)})
  }

  handleMovePiece(piece, fromSquare, toSquare) {
    const newPieces = this.state.pieces
      .map((curr, index) => {
        if (piece.index === index) {
          return `${piece.name}@${toSquare}`
        } else if (curr.indexOf(toSquare) === 2) {
          return false // To be removed from the board
        }
        return curr
      })
      .filter(Boolean)

    this.setState({pieces: newPieces, selectedPiece: null})
  }

  handleSelectPiece(selectedPiece) {
    this.setState({selectedPiece})
  }

  render() {
    const {pieces} = this.state
    return (
      <div className="demo">
        <Chess
          pieces={pieces}
          onMovePiece={this.handleMovePiece}
          onSelectPiece={this.handleSelectPiece}
        />
      </div>
    )
  }
}

module.exports = Demo
